metrics/src/main/java/io/smallrye/graphql/metrics/MetricsInterceptor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 IBM Corporation and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.smallrye.graphql.metrics;
+
+import java.lang.reflect.Method;
+
+import javax.annotation.Priority;
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.MetricRegistry.Type;
+import org.eclipse.microprofile.metrics.SimpleTimer;
+import org.eclipse.microprofile.metrics.SimpleTimer.Context;
+import org.eclipse.microprofile.metrics.annotation.RegistryType;
+import org.jboss.logging.Logger;
+
+@Dependent
+@GraphQLApi
```

Comment:
It's not clear to me how you expect the interceptors to get bound to the relevant methods.
And what is this ` @GraphQLApi` annotation for? From first look I thought it might be an interceptor binding, but it's not.

Reply:
Sorry - I tested this in OpenLiberty with the following code used to bind the interceptors:
```
AnnotatedType<GraphQLApi> bindingType = beanManager.createAnnotatedType(GraphQLApi.class);
        beforeBeanDiscovery.addInterceptorBinding(bindingType);

        AnnotatedType<GraphQLMetricsInterceptor> metricsInterceptorType = beanManager.createAnnotatedType(GraphQLMetricsInterceptor.class);
        beforeBeanDiscovery.addAnnotatedType(metricsInterceptorType, CDIServiceUtils.getAnnotatedTypeIdentifier(metricsInterceptorType, this.getClass()));
```

In OL's case, I extended the interceptor so that I could programmatically specify the MetricRegistry.  But I could add this code in a CDI extension that automatically registers it if the module is packaged in the application.  I could also add some sort of `register` method if we wanted users to "opt-in" even if the module is packaged with the app.  I lean toward automatic registration.  Wdyt?

Reply:
Including a CDI extension for registering the interceptor would definitely be +1 if we go with the interceptor approach.

However, with that code you would apply the interceptor to *all* methods of the class, we need to detect just the methods with `@Query` or `@Mutation`, so maybe we should use these two annotations as bindings instead?

Registering metrics eagerly at boot would be a good thing too. The CDI extension could scan the annotated types, locate `@Query` and `@Mutation` methods and register metrics for them.